datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -101,20 +158,127 @@ func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
 	return nil
 }
 
+// insertSegment inserts the given segment to the given buffer.
+// Assumption:
+//  1. it is guaranteed that the order of segments is correct.
+//  2. the inserted messages may be duplicated.
+//
+// We can consider the following cases.
+// case1. new segment with index 0 is inserted into newly created segment slice.
+// case2. new consecutive segment is inserted into the right position.
+// case3. duplicated segment is ignored.
+// case4. new sparse segment shouldn't be given, so return an error.
+func insertSegment(segment *Segment, buffer [][]*Segment) ([][]*Segment, error) {
+	for idx, segments := range buffer {
+		length := len(segments)
+		if length > 0 && segments[0].key == segment.key {
+			// there is a missing segment which should not exist.
+			if segment.index > uint64(length) {
+				return buffer, errors.New(missingSegmentErrorMsg)
+			}
+
+			// the segment is already inserted to buffer.
+			if segment.index < uint64(length) {
+				logger.Warn("the message is duplicated", "segment", segment)
+				return buffer, nil
+			}
+
+			// insert the segment to the buffer.
+			buffer[idx] = append(segments, segment)
+			return buffer, nil
+		}
+	}
+
+	if segment.index == 0 {
+		// create a segment slice and append it.
+		buffer = append(buffer, []*Segment{segment})
```

Comment:
I think it could be more efficient if these are placed before the `for` loop above.

Reply:
If a segment with index 0 is duplicated, then it will be handled in the above for loop.